kroxylicious-kms-provider-aws-kms/DEV_GUIDE.md

```diff
@@ -0,0 +1,81 @@
+Here's some information that might be helpful when developing this module.
+
+# Testing the ability to authenticate with EC2 metadata
+
+In order to test this feature, you need to deploy Kroxylicious to an EC2 instance. The 
+EC2 instance needs to be associated with a properly configured IAM role. The EC2 credentials
+provider will use that to authenticate.
+
+Here's some instructions that will help you test end to end.  For simplicity, these instructions
+co-locate a Kafka Broker on the same instance.  Of course, you can choose to run the Kafka
+Cluster in any network accessible location.  These instructions also assume you want to run
+your Kafka client off-EC2.
+
+1. Create an EC2 instance on AWS - `t2.micro` instance will suffice.
+2. Enable SSH with key pair. The remainder of the instructions assume you've download the key into `.ssh/MyEC2KeyPair.pem`.
+3. Enable inbound traffic to port 22 (for SSH) and 9192 - 9193 (for Kafka traffic).
+4. Follow the 'Authenticating using AWS EC2 metadata' instructions in https://kroxylicious.io/kroxylicious/#assembly-aws-kms-proxy
+   to create the policy, IAM role and assign the IAM role to the EC2 host.
+5. SSH into the EC2 instance `ssh  -i "~/.ssh/MyEC2KeyPair.pem" ec2-xxx-xxx-xxx-xxx.compute-1.amazonaws.com` and install packages ready to run docker containers and kroxylicious.
+   ```bash
+    sudo yum install -y java maven docker
+    sudo service docker start
+    sudo usermod -a -G docker ec2-user
+   ```
+6. Logout/login to EC2 again to pick up the new group.
+7. Run Kafka on the EC2 instance (inside a docker container)
+   ```bash
+   docker run -d -p 9092:9092 --name broker apache/kafka:latest
+   ```
+8. Scp a Kroxylicious dist to the EC2 instance.
+   ```bash
+   scp  -i "~/.ssh/MyEC2KeyPair.pem" /kroxylicious-app-0.9.0-SNAPSHOT-bin.zip ec2-user@ec2-xxx-xxx-xxx-xxx.compute-1.amazonaws.com:.
+   ```
+9. Back on the EC2 instance, unzip Kroxylicious
+   ```bash
+   unzip kroxylicious-app-0.9.0-SNAPSHOT-bin.zip
+   ```
+10. Create a config file for Kroxylicious updating the virtual clusters bootstrap address to the public address of the EC2 instance.
+    ```yaml
+     virtualClusters:
+     - demo:
+       targetCluster:
+         bootstrap_servers: localhost:9092
+       clusterNetworkAddressConfigProvider:
+         type: PortPerBrokerClusterNetworkAddressConfigProvider
+         config:
+           bootstrapAddress: ec2-xx-xx-xx-xx.compute-1.amazonaws.com:9192
+       logNetwork: false
+       logFrames: false
+     filters:
+     - type: RecordEncryption
+       config:
+         kms: AwsKmsService
+         kmsConfig:
+           region: us-east-1
+           endpointUrl: https://kms.us-east-1.amazonaws.com
+         credentialsProvider:
+           type: ec2Metadata
+           iamRole: KroxyliciousInstance
```

Comment:
I found this a bit tricky to debug, since my role was named differently. The symptom was logs like:

```
java.util.concurrent.CompletionException: io.kroxylicious.kms.service.KmsException: resolveAlias failed after 3 attempts, last failure message: io.kroxylicious.kms.service.KmsException: Operation failed, HTTP status code 404, response: <?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
		 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>404 - Not Found</title>
 </head>
 <body>
  <h1>404 - Not Found</h1>
 </body>
</html>

	at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(CompletableFuture.java:323) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:412) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:421) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniComposeExceptionally.tryFire(CompletableFuture.java:1086) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2238) ~[?:?]
	at io.kroxylicious.filter.encryption.kms.ResilientKms.lambda$schedule$4(ResilientKms.java:119) ~[kroxylicious-record-encryption-0.9.0-SNAPSHOT.jar:?]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2238) ~[?:?]
	at io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProvider.propagateResultToFuture(Ec2MetadataCredentialsProvider.java:202) ~[kroxylicious-kms-provider-aws-kms-0.9.0-SNAPSHOT.jar:?]
	at io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProvider.lambda$refreshCredential$7(Ec2MetadataCredentialsProvider.java:194) ~[kroxylicious-kms-provider-aws-kms-0.9.0-SNAPSHOT.jar:?]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:658) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:888) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:526) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1575) [?:?]
```

Which is quite a bit to parse and see that's it's token-request-related not KMS-request-related (if that makes sense). Maybe the OperationFailed KmsException should describe which operation has failed.

Reply:
Good point.  I'm surprised by the HTML response too.

Reply:
The endpoint always gives a HTML response under 404 conditions, even if you give it the Accept header.

Reply:
I've included the uri in the exception message.